packages/protocol/contracts/L1/ProverPool.sol

```diff
@@ -0,0 +1,364 @@
+        if (amount == 0) {
+            if (rewardPerGas != 0 || maxCapacity != 0) {
+                revert INVALID_PARAMS();
+            }
+        } else {
+            _stake(msg.sender, amount, rewardPerGas, maxCapacity);
+        }
+    }
+
+    function exit() external nonReentrant {
+        _withdraw(msg.sender);
+        _exit(msg.sender);
+    }
+
+    // Withdraws staked tokens back from matured an exit
+    function withdraw() external nonReentrant {
+        if (!_withdraw(msg.sender)) revert NO_MATURE_EXIT();
+    }
+
+    // Returns a staker's information
+    function getStaker(address addr)
+        public
+        view
+        returns (Staker memory staker, Prover memory prover)
+    {
+        staker = stakers[addr];
+        if (staker.proverId != 0) {
+            unchecked {
+                prover = provers[staker.proverId];
+            }
+        }
+    }
+
+    // Returns the pool's current total capacity
+    function getCapacity() public view returns (uint256 capacity) {
+        unchecked {
+            for (uint256 i = 1; i <= MAX_NUM_PROVERS; ++i) {
+                capacity += provers[i].currentCapacity;
+            }
+        }
+    }
+
+    function getProvers()
+        public
+        view
+        returns (Prover[] memory _provers, address[] memory _stakers)
+    {
+        _provers = new Prover[](MAX_NUM_PROVERS);
+        _stakers = new address[](MAX_NUM_PROVERS);
+        for (uint256 i; i < MAX_NUM_PROVERS; ++i) {
+            _provers[i] = provers[i + 1];
+            _stakers[i] = idToProver[i + 1];
+        }
+    }
+
+    function _stake(
+        address addr,
+        uint64 amount,
+        uint16 rewardPerGas,
+        uint16 maxCapacity
+    )
+        private
+    {
+        // Check parameters
+        if (
+            maxCapacity < MAX_CAPACITY_LOWER_BOUND
+                || amount / maxCapacity < MIN_STAKE_PER_CAPACITY
+                || rewardPerGas == 0
+        ) revert INVALID_PARAMS();
+
+        // Reuse tokens that are exiting
+        Staker storage staker = stakers[addr];
+
+        unchecked {
+            if (staker.exitAmount >= amount) {
+                staker.exitAmount -= amount;
+            } else {
+                uint64 burnAmount = (amount - staker.exitAmount);
+                TaikoToken(resolve("taiko_token", false)).burn(addr, burnAmount);
+                staker.exitAmount = 0;
+            }
+        }
+
+        staker.exitRequestedAt =
+            staker.exitAmount == 0 ? 0 : uint64(block.timestamp);
+
+        staker.maxCapacity = maxCapacity;
+
+        // Find the prover id
+        uint8 proverId = 1;
+        for (uint8 i = 2; i <= MAX_NUM_PROVERS;) {
+            if (provers[proverId].stakedAmount > provers[i].stakedAmount) {
+                proverId = i;
+            }
+            unchecked {
+                ++i;
+            }
+        }
+
+        if (provers[proverId].stakedAmount >= amount) {
+            revert PROVER_NOT_GOOD_ENOUGH();
```

Comment:
If it's not good enough, can we have the amount not burned so that the staker can withdraw soon?